### PR TITLE
Tenant login and register handlers

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This adds initial Tenant handlers for Login and Register which are mostly just passthrough handlers to the Quarterdeck endpoints. There is currently no Tenant user checks or resource creation which we will have to do for new users.

Fixes SC-12798

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

This review should be fairly straightforward, the main thing is capturing the follow-up work in terms of stories.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?

### CODE REVIEW

File: pkg/tenant/api/v1/api.go [hunk 1]
```diff
@@ -75,6 +78,29 @@ type StatusReply struct {
...
+	Email    string `json:"email"`
+	Password string `json:"password"`
+	PwCheck  string `json:"pwcheck"`
+}
+
+type RegisterReply struct {
+	Email   string `json:"email"`
+	Message string `json:"message"`
+	Role    string `json:"role"`
+}
```
Comment (bbengfort) [inline pkg/tenant/api/v1/api.go:92]: Just an FYI that we will have to update this after tomorrow's feature session - I'm aware that we aren't collecting all of the registration data but needed to catch up with the front end team to find out what data was coming in.

When we create the story to update the register requests structs it should update for _both_ tenant and Quarterdeck.

File: pkg/tenant/api/v1/client.go [hunk 2]
```diff
@@ -101,6 +101,32 @@ func (s *APIv1) SignUp(ctx context.Context, in *ContactInfo) (err error) {
...
+	}
+
+	out = &RegisterReply{}
+	if _, err = s.Do(req, out, true); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (s *APIv1) Login(ctx context.Context, in *LoginRequest) (out *AuthReply, err error) {
```
Comment (bbengfort) [inline pkg/tenant/api/v1/client.go:117]: Let's create a story for also adding `Refresh` here which Prema implemented today; sorry I forgot to add it to this story, but I don't want to block the completion of this PR.

File: pkg/tenant/api/v1/client_test.go [hunk 3]
```diff
@@ -132,6 +132,71 @@ func TestStatus(t *testing.T) {
...
+	// Create a client to execute tests against the test server
+	client, err := api.New(ts.URL)
+	require.NoError(t, err, "could not create client")
+
+	// Create a new register request
+	req := &api.RegisterRequest{
+		Name:     "Leopold Wentzel",
+		Email:    "leopold.wentzel@gmail.com",
+		Password: "hunter2",
+		PwCheck:  "hunter2",
```
Comment (bbengfort) [inline pkg/tenant/api/v1/client_test.go:162]: :-)

File: pkg/tenant/auth.go [hunk 4]
```diff
@@ -13,6 +14,115 @@ import (
...
+		return
+	}
+
+	// Simple validation of the provided password
+	// Note: Quarterdeck also checks this along with password strength, but this allows
+	// us to filter some bad requests before they reach Quarterdeck.
+	if params.Password != params.PwCheck {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("passwords do not match"))
+		return
+	}
```
Comment (bbengfort) [inline pkg/tenant/auth.go:44]: Excellent! Thank you!

File: pkg/tenant/auth.go [hunk 5]
```diff
@@ -13,6 +14,115 @@ import (
...
+
+	// TODO: Handle error status codes returned by Quarterdeck
+	var reply *qd.RegisterReply
+	if reply, err = s.quarterdeck.Register(c.Request.Context(), req); err != nil {
+		log.Error().Err(err).Msg("could not register user")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not complete registration"))
+		return
+	}
+
+	// TODO: Send verification email to the provided email address
```
Comment (bbengfort) [inline pkg/tenant/auth.go:62]: We'll have to discuss who's responsibility this is - Quarterdeck or Tenant?

Comment (bbengfort) [inline pkg/tenant/auth.go:62]:
```suggestion
	// TODO: Send verification email to the provided email address
	// TODO: Create tenant and other Tenant-specific resources for the user
```

File: pkg/tenant/auth.go [hunk 6]
```diff
@@ -13,6 +14,115 @@ import (
...
+
+	// Make the register request to Quarterdeck
+	req := &qd.RegisterRequest{
+		Name:     params.Name,
+		Email:    params.Email,
+		Password: params.Password,
+		PwCheck:  params.PwCheck,
+	}
+
+	// TODO: Handle error status codes returned by Quarterdeck
```
Comment (bbengfort) [inline pkg/tenant/auth.go:54]: The quarterdeck client returns a `StatusError` struct that can be used to get the status code from:

https://github.com/rotationalio/ensign/blob/main/pkg/quarterdeck/api/v1/errors.go#L103

Let's create a story for this TODO and mention the use of this error struct for better response handling.

File: pkg/tenant/auth.go [hunk 7]
```diff
@@ -13,6 +14,115 @@ import (
...
+	// TODO: Add user state checks and create required resources for first logins
+	// (tenants, projects, etc.)
+
+	// Protect the frontend from CSRF attacks by setting the double cookie tokens
+	expiresAt := time.Now().Add(doubleCookiesMaxAge)
+	if err := middleware.SetDoubleCookieToken(c, s.conf.Auth.CookieDomain, expiresAt); err != nil {
+		log.Error().Err(err).Msg("could not set cookies on login reply")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not set cookies"))
+		return
+	}
```
Comment (bbengfort) [inline pkg/tenant/auth.go:116]: Thank you for adding this!

**Review by bbengfort — APPROVED** [review]
Looks good!

File: pkg/tenant/auth.go (hunk 5 above)
Comment (pdeziel) [reply pkg/tenant/auth.go:62]: Hmm good question

**Pull request merged**